core/rio/turtle/src/main/java/org/eclipse/rdf4j/rio/turtle/TurtleWriter.java

```diff
@@ -701,4 +735,109 @@ private void wrapLine(boolean space) throws IOException {
+				Model contextData = bufferedStatements.filter(null, null, null, context);
+				Set<Resource> processedSubjects = new HashSet<>();
+				Optional<Resource> nextSubject = nextSubject(contextData, processedSubjects);
+				while (nextSubject.isPresent()) {
+					processSubject(contextData, nextSubject.get(), processedSubjects);
+					nextSubject = nextSubject(contextData, processedSubjects);
+				}
+			}
+		} else {
+			// group by subject
+
+			Set<Resource> processedSubjects = new HashSet<>();
+			Optional<Resource> nextSubject = nextSubject(bufferedStatements, processedSubjects);
+			while (nextSubject.isPresent()) {
+				processSubject(bufferedStatements, nextSubject.get(), processedSubjects);
+				nextSubject = nextSubject(bufferedStatements, processedSubjects);
+			}
+		}
+		bufferedStatements.clear();
+	}
+
+	private Optional<Resource> nextSubject(Model contextData, Set<Resource> processedSubjects) {
+		for (Resource subject : contextData.subjects()) {
+			if (processedSubjects.contains(subject)) {
+				continue;
+			}
+			if (subject instanceof BNode && inlineBNodes) {
+				Set<Resource> otherSubjects = contextData.filter(null, null, subject).subjects();
+				if (otherSubjects.stream().anyMatch(s -> !processedSubjects.contains(s))) {
+					continue;
+				}
+			}
+			return Optional.of(subject);
+		}
+		return Optional.empty();
+	}
+
+	private void processSubject(Model contextData, Resource subject, Set<Resource> processedSubjects) {
+		if (processedSubjects.contains(subject)) {
+			return;
+		}
+
+		Set<IRI> processedPredicates = new HashSet<>();
+
+		// give rdf:type preference over other predicates.
+		for (Statement typeStatement : contextData.getStatements(subject, RDF.TYPE, null)) {
+			boolean canInlineObject = inlineValue(contextData, typeStatement.getObject());
+			handleStatementInternal(typeStatement, false,
+					inlineValue(contextData, typeStatement.getSubject()),
+					canInlineObject);
+
+			if (canInlineObject && typeStatement.getObject() instanceof BNode) {
+				processSubject(contextData, (BNode) typeStatement.getObject(), processedSubjects);
+			}
+		}
+
+		processedPredicates.add(RDF.TYPE);
+
+		// retrieve other statement from this context with the same
+		// subject, and output them grouped by predicate
+		for (Statement subjectStatement : contextData.getStatements(subject, null, null)) {
+			IRI predicate = subjectStatement.getPredicate();
+			if (!processedPredicates.contains(predicate)) {
+				for (Statement toWrite : contextData.getStatements(subject, predicate, null)) {
+					boolean canInlineObject = inlineValue(contextData, toWrite.getObject());
+					handleStatementInternal(toWrite, false, inlineValue(contextData, toWrite.getSubject()),
+							canInlineObject);
+					if (canInlineObject && toWrite.getObject() instanceof BNode) {
+						processSubject(contextData, (BNode) toWrite.getObject(), processedSubjects);
+					}
+				}
+				processedPredicates.add(predicate);
+			}
+
+		}
```

Comment:
Since these pieces of code are so similar I'm wondering if maybe you would prefer to use the java 8 streams grouping collector (https://www.baeldung.com/java-groupingby-collector) and then sort it so that `rdf:type` gets preference. 

A more performant alternative would be to sort `contextData.getStatements(subject, null, null)` by the predicate with preference to `rdf:type` and then grouping could be done by looking at the previous predicate.

Reply:
I'll take a look at DRYing this up a bit.

Reply:
I had a quick look at a streaming/sorting approach, but to be honest I'm not convinced I can easily come up with an alternative that is both more performant and also still easy to read/maintain. Not really worth investing too much time in right now I think.